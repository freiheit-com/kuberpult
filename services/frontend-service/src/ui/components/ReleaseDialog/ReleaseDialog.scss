.release-dialog {

  div.MuiPaper-root {
    background-color: transparent;
    border-radius: $release-dialog-outer-border-radius;
    overflow-y: inherit;
  }
  .content-area {
    display: flex;
  }
  .content-right {
    margin-right: 20px;
    margin-left: auto;
    display: flex;
  }

  .release-dialog-app-bar {
    border-radius: $release-dialog-top-border-radius;
    background: var(--mdc-theme-primary);
    color: var(--mdc-theme-on-primary);
    display: flex;
    flex-direction: row;
    justify-content: space-between;

    .release-dialog-app-bar-data {
      margin: 24px 10px 24px 47px;
      display: flex;
      flex-direction: column;
      justify-content: space-between;

      .release-dialog-message {
        @extend .sub-headline1;
        display: flex;
        justify-content: space-between;

        .release-dialog-commitMessage {
          text-overflow: wrap;
        }
      }
    }

    .release-dialog-commitId {
      height: 28px;
      white-space: nowrap;
      width: auto;
      padding: 2px 6px;
      border: 1px solid var(--mdc-theme-on-primary);
      border-radius: $release-dialog-inner-border-radius;
      margin-top: 26px;
      margin-bottom: auto;
      text-align: center;
      @extend .text-regular;
    }

    button.release-dialog-close {
      display: flex;
      margin: 20px 10px 100px;
    }
  }

  .release-env-group-list {
    background-color: var(--mdc-theme-surface);
    display: flex;
    flex-direction: column;
    overflow: overlay;
    gap: 22px;
    margin: 0px;
    padding: 35px 0px;
    align-items: center;
    border-radius: $release-dialog-bottom-border-radius;

    .release-env-list {
        padding-left: 0;
        margin-bottom: 0;
    }

    .env-card {
      background-color: var(--mdc-theme-background);
      display: flex;
      flex-direction: column;
      width: 900px;
      padding: 0px;
      margin-top: 10px;
      margin-bottom: 0px;
      margin-left: 20px;
      margin-right: 20px;
      border-radius: $release-dialog-inner-border-radius;
      gap: 16px;

      .env-card-header {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        .release-environment {
          top: $release-dialog-label-displacement;
          left: $release-dialog-label-displacement;
          height: 42px;
          .env-card-label {
            display: flex;

            svg.env-card-env-lock-icon {
              width: $release-dialog-lock-icon-size;
              height: $release-dialog-lock-icon-size;
              margin: unset;
            }

            .env-card-env-locks:first-child {
              margin-left: 55px;
              display: flex;
              flex-direction: row;
            }

          }
        }

        .env-card-app-locks {
          display: flex;
          position: relative;
          top: 20px;
          right: 36px;
          flex-direction: row;
          gap: 13px;
        }
      }

      .env-card-data {
        display: flex;
        flex-direction: row;
        padding-left: 27px;
        @extend .text-medium;
      }

      .env-card-buttons {
        margin-left: auto;
        margin-right: 18px;
        margin-bottom: 17px;
        display: flex;
        flex-direction: row;
        gap: 20px;
        button {
          padding: 8px 19px;
        }

        .env-card-add-lock-btn {
          border: 1px solid var(--mdc-theme-primary);
          border-radius: $release-dialog-inner-border-radius;
          @extend .text-bold;
          .icon {
            margin-right: 9px;
          }
        }
      }
    }
  }
}

button.button-lock {
  min-width: unset;
  padding: unset;
  height: auto;
}

button.mdc-button.env-card-deploy-btn {
  background-color: var(--mdc-theme-primary);
  color: var(--mdc-theme-on-primary);
  border-radius: $release-dialog-inner-border-radius;
  @extend .text-bold;
}

button.mdc-button.env-card-deploy-btn.btn-disabled {
  background-color: var(--mdc-theme-on-surface);
}