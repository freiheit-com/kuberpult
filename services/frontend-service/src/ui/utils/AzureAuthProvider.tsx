/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright 2023 freiheit.com*/
import * as React from 'react';
import { BrowserHeaders } from 'browser-headers';
import { Configuration, PublicClientApplication } from '@azure/msal-browser';
import {
    MsalProvider,
    AuthenticatedTemplate,
    useIsAuthenticated,
    UnauthenticatedTemplate,
    useMsal,
} from '@azure/msal-react';
import { GetFrontendConfigResponse } from '../../api/api';
import { createStore } from 'react-use-sub';
import { grpc } from '@improbable-eng/grpc-web';

type FrontendConfig = {
    configs: GetFrontendConfigResponse;
    configReady: boolean;
};

export const [useFrontendConfig, UpdateFrontendConfig] = createStore<FrontendConfig>({
    configs: {
        sourceRepoUrl: '',
        kuberpultVersion: '0',
    },
    configReady: false,
});

type AzureAuthSubType = {
    authHeader: grpc.Metadata & {
        Authorization?: String;
    };
    authReady: boolean;
};

export const [useAzureAuthSub, AzureAuthSub] = createStore<AzureAuthSubType>({
    authHeader: new BrowserHeaders({}),
    authReady: false,
});

const getMsalConfig = (configs: GetFrontendConfigResponse): Configuration => ({
    auth: {
        clientId: configs.authConfig?.azureAuth?.clientId || '',
        authority: `${configs.authConfig?.azureAuth?.cloudInstance || ''}${
            configs.authConfig?.azureAuth?.tenantId || ''
        }`,
        redirectUri: configs.authConfig?.azureAuth?.redirectURL || '',
    },
    cache: {
        cacheLocation: 'sessionStorage',
        storeAuthStateInCookie: false,
    },
});

// - Email scope was added later so kuberpult can extract the email from requests (the author)
//   and send it along to the backend
const loginRequest = {
    scopes: ['email'],
};

export const AcquireToken: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    const { instance, accounts } = useMsal();

    React.useEffect(() => {
        const request = {
            ...loginRequest,
            account: accounts[0],
        };
        instance
            .acquireTokenSilent(request)
            .then((response) => {
                AzureAuthSub.set({
                    authHeader: new BrowserHeaders({ Authorization: response.idToken }),
                    authReady: true,
                });
            })
            .catch(() => {
                instance.acquireTokenPopup(request).then((response) => {
                    AzureAuthSub.set({
                        authHeader: new BrowserHeaders({ Authorization: response.idToken }),
                        authReady: true,
                    });
                });
            });
    }, [instance, accounts]);

    return <>{children}</>;
};

export const AzureAutoSignIn = (): JSX.Element => {
    const isAuthenticated = useIsAuthenticated();
    const { instance } = useMsal();
    React.useEffect(() => {
        if (!isAuthenticated) {
            instance.loginRedirect(loginRequest);
        }
    }, [instance, isAuthenticated]);
    return <>Redirecting to login</>;
};

export const AzureAuthProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    const { configs, configReady } = useFrontendConfig((c) => c);
    const msalInstance = React.useMemo(() => new PublicClientApplication(getMsalConfig(configs)), [configs]);
    if (!configReady) return null;

    const useAzureAuth = configs.authConfig?.azureAuth?.enabled;
    if (!useAzureAuth) {
        AzureAuthSub.set({ authReady: true });
        return <>{children}</>;
    }

    return (
        <MsalProvider instance={msalInstance}>
            <AuthenticatedTemplate>
                <AcquireToken>{children}</AcquireToken>
            </AuthenticatedTemplate>
            <UnauthenticatedTemplate>
                <AzureAutoSignIn />
            </UnauthenticatedTemplate>
        </MsalProvider>
    );
};
