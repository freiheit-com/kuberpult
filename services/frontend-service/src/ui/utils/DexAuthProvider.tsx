/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright freiheit.com*/
import { jwtDecode } from 'jwt-decode';
import { TopAppBar } from '../components/TopAppBar/TopAppBar';
import { Button } from '../components/button';
import React from 'react';

export const LoginPage: React.FC = () => {
    // Redirect the user to the Dex Login
    const handleRedirect = React.useCallback(() => {
        // A random value is added to avoid browser cashing when redirecting to DEX.
        const id = window.crypto.randomUUID();
        window.location.href = `/login?random_value=${id}`;
    }, []);

    return (
        <div>
            <TopAppBar showAppFilter={false} showTeamFilter={false} showWarningFilter={false} />
            <main className="main-content">
                <h1 className="environment_name">{'Log in to Dex'}</h1>
                <h3 className="page_description">{'You are currently not logged in. Please log in to continue.'}</h3>
                <div className="space_apart_row">
                    <Button
                        label={'Login'}
                        className={'button-main env-card-deploy-btn mdc-button--unelevated'}
                        onClick={handleRedirect}
                        highlightEffect={false}
                    />
                </div>
            </main>
        </div>
    );
};

// Validates the token expiring date
export function isTokenValid(): boolean {
    const cookieValue = document.cookie
        .split('; ')
        .find((row) => row.startsWith('kuberpult.oauth='))
        ?.split('=')[1];
    if (!cookieValue) return false;

    const decodedToken = jwtDecode(cookieValue);
    if (!decodedToken) return false;

    const currentTime = Date.now() / 1000;
    if (!decodedToken.exp) return false;
    if (decodedToken.exp < currentTime) {
        return false;
    }
    return true;
}
