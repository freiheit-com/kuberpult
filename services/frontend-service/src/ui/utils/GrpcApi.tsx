/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright 2023 freiheit.com*/
import * as api from '../../api/api';

export interface Api {
    // overview service is used to get data about apps, envs and releases
    overviewService(): api.OverviewService;
    // batch service can make changes, e.g. adding/removing locks, deploying..
    batchService(): api.BatchService;
    // config service returns basics configuration like for azure auth
    configService(): api.FrontendConfigService;
}

class GrpcApi implements Api {
    _overviewService: api.OverviewService;
    _batchService: api.BatchService;
    _configService: api.FrontendConfigService;
    constructor() {
        // eslint-disable-next-line no-restricted-globals
        const gcli = new api.GrpcWebImpl(location.protocol + '//' + location.host, {});
        this._overviewService = new api.OverviewServiceClientImpl(gcli);
        this._batchService = new api.BatchServiceClientImpl(gcli);
        this._configService = new api.FrontendConfigServiceClientImpl(gcli);
    }
    overviewService(): api.OverviewService {
        return this._overviewService;
    }
    batchService(): api.BatchService {
        return this._batchService;
    }
    configService(): api.FrontendConfigService {
        return this._configService;
    }
}

export const useApi = new GrpcApi();
