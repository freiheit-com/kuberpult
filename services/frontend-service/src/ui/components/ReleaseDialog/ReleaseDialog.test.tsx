/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright 2023 freiheit.com*/
import { ReleaseDialog, ReleaseDialogProps } from './ReleaseDialog';
import { render } from '@testing-library/react';
import { UpdateOverview, updateReleaseDialog } from '../../utils/store';
import { Priority, Release } from '../../../api/api';

describe('Release Dialog', () => {
    interface dataT {
        name: string;
        props: ReleaseDialogProps;
        rels: Release[];
        expect_message: boolean;
        expect_queues: number;
        data_length: number;
    }
    const data: dataT[] = [
        {
            name: 'normal release',
            props: {
                app: 'test1',
                version: 2,
                release: {
                    version: 2,
                    sourceMessage: 'test1',
                    sourceAuthor: 'test',
                    sourceCommitId: 'commit',
                    createdAt: new Date(2002),
                    undeployVersion: false,
                    prNumber: '#1337',
                },
                envs: [
                    {
                        name: 'prod',
                        locks: { envLock: { message: 'envLock', lockId: 'ui-envlock' } },
                        applications: {
                            test1: {
                                name: 'test1',
                                version: 2,
                                locks: { applock: { message: 'appLock', lockId: 'ui-applock' } },
                                queuedVersion: 0,
                                undeployVersion: false,
                            },
                        },
                        distanceToUpstream: 0,
                        priority: Priority.UPSTREAM,
                    },
                ],
            },
            rels: [],

            expect_message: true,
            expect_queues: 0,
            data_length: 1,
        },
        {
            name: 'two envs release',
            props: {
                app: 'test1',
                version: 2,
                release: {
                    version: 2,
                    sourceMessage: 'test1',
                    sourceAuthor: 'test',
                    sourceCommitId: 'commit',
                    createdAt: new Date(2002),
                    undeployVersion: false,
                    prNumber: '#1337',
                },
                envs: [
                    {
                        name: 'prod',
                        locks: { envLock: { message: 'envLock', lockId: 'ui-envlock' } },
                        applications: {
                            test1: {
                                name: 'test1',
                                version: 2,
                                locks: { applock: { message: 'appLock', lockId: 'ui-applock' } },
                                queuedVersion: 0,
                                undeployVersion: false,
                            },
                        },
                        distanceToUpstream: 0,
                        priority: Priority.UPSTREAM,
                    },
                    {
                        name: 'dev',
                        locks: { envLock: { message: 'envLock', lockId: 'ui-envlock' } },
                        applications: {
                            test1: {
                                name: 'test1',
                                version: 3,
                                locks: { applock: { message: 'appLock', lockId: 'ui-applock' } },
                                queuedVersion: 666,
                                undeployVersion: false,
                            },
                        },
                        distanceToUpstream: 0,
                        priority: Priority.UPSTREAM,
                    },
                ],
            },
            rels: [],

            expect_message: true,
            expect_queues: 1,
            data_length: 2,
        },
        {
            name: 'no release',
            props: {
                app: 'test1',
                version: -1,
                release: {} as Release,
                envs: [],
            },
            rels: [],
            expect_message: false,
            expect_queues: 0,
            data_length: 0,
        },
    ];

    describe.each(data)(`Renders a Release Dialog`, (testcase) => {
        it(testcase.name, () => {
            // when
            UpdateOverview.set({
                applications: { [testcase.props.app as string]: { releases: testcase.rels } },
                environments: testcase.props.envs,
                environmentGroups: [
                    {
                        environmentGroupName: 'dev',
                        environments: testcase.props.envs,
                        distanceToUpstream: 2,
                    },
                ],
            } as any);
            updateReleaseDialog(testcase.props.app, testcase.props.version);
            render(<ReleaseDialog {...testcase.props} />);
            if (testcase.expect_message) {
                expect(document.querySelector('.release-dialog-message')?.textContent).toContain(
                    testcase.props.release.sourceMessage
                );
            } else {
                expect(document.querySelector('.release-dialog-message') === undefined);
            }
            expect(document.querySelectorAll('.env-card-data')).toHaveLength(testcase.data_length);
        });
    });

    describe.each(data)(`Renders the environment cards`, (testcase) => {
        it(testcase.name, () => {
            // when
            UpdateOverview.set({
                applications: { [testcase.props.app as string]: { releases: testcase.rels } },
                environments: testcase.props.envs,
                environmentGroups: [
                    {
                        environmentGroupName: 'dev',
                        environments: testcase.props.envs,
                        distanceToUpstream: 2,
                    },
                ],
            } as any);
            updateReleaseDialog(testcase.props.app, testcase.props.version);
            render(<ReleaseDialog {...testcase.props} />);
            expect(document.querySelector('.release-env-list')?.children).toHaveLength(testcase.props.envs.length);
        });
    });

    describe.each(data)(`Renders the environment locks`, (testcase) => {
        it(testcase.name, () => {
            // when
            UpdateOverview.set({
                applications: { [testcase.props.app as string]: { releases: testcase.rels } },
                environments: testcase.props.envs,
                environmentGroups: [
                    {
                        environmentGroupName: 'dev',
                        environments: testcase.props.envs,
                        distanceToUpstream: 2,
                    },
                ],
            } as any);
            updateReleaseDialog(testcase.props.app, testcase.props.version);
            render(<ReleaseDialog {...testcase.props} />);
            expect(document.body).toMatchSnapshot();
            expect(document.querySelectorAll('.release-env-group-list')).toHaveLength(1);

            testcase.props.envs.forEach((env) => {
                expect(document.querySelector('.env-locks')?.children).toHaveLength(Object.values(env.locks).length);
            });
        });
    });

    describe.each(data)(`Renders the queuedVersion`, (testcase) => {
        it(testcase.name, () => {
            // when
            UpdateOverview.set({
                applications: { [testcase.props.app as string]: { releases: testcase.rels } },
                environments: testcase.props.envs,
                environmentGroups: [
                    {
                        environmentGroupName: 'dev',
                        environments: testcase.props.envs,
                        distanceToUpstream: 2,
                    },
                ],
            } as any);
            updateReleaseDialog(testcase.props.app, testcase.props.version);
            render(<ReleaseDialog {...testcase.props} />);
            expect(document.querySelectorAll('.env-card-data-queue')).toHaveLength(testcase.expect_queues);
        });
    });
});
