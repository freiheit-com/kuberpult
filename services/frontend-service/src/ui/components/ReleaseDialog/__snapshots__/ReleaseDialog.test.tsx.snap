// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Release Dialog Renders the environment locks no release 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div>
      <div />
    </div>
  </div>
  <div
    class="MuiModal-root MuiDialog-root release-dialog css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthMd MuiDialog-paperFullWidth css-142erj9-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <div
          class="release-dialog-app-bar"
        >
          <div
            class="release-dialog-app-bar-data"
          >
            <div
              class="release-dialog-message"
            >
              <span
                class="release-dialog-commitMessage"
              />
            </div>
            <div
              class="release-dialog-createdAt"
            />
            <div
              class="release-dialog-author"
            />
          </div>
          <span
            class="release-dialog-commitId"
            title="Commit Hash of the source repository."
          />
          <button
            aria-label=""
            class="mdc-button release-dialog-close"
          >
            <div
              class="mdc-button__ripple"
            />
            <svg>
              Close.svg
            </svg>
          </button>
        </div>
        <div
          class="release-env-group-list"
        >
          <ul
            class="release-env-list"
          />
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`Release Dialog Renders the environment locks normal release 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div>
      <div />
    </div>
  </div>
  <div
    class="MuiModal-root MuiDialog-root release-dialog css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthMd MuiDialog-paperFullWidth css-142erj9-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <div
          class="release-dialog-app-bar"
        >
          <div
            class="release-dialog-app-bar-data"
          >
            <div
              class="release-dialog-message"
            >
              <span
                class="release-dialog-commitMessage"
              >
                test1
              </span>
            </div>
            <div
              class="release-dialog-createdAt"
            >
              <div>
                some formatted date
              </div>
            </div>
            <div
              class="release-dialog-author"
            >
              Author: test
            </div>
          </div>
          <span
            class="release-dialog-commitId"
            title="Commit Hash of the source repository."
          >
            commit
          </span>
          <button
            aria-label=""
            class="mdc-button release-dialog-close"
          >
            <div
              class="mdc-button__ripple"
            />
            <svg>
              Close.svg
            </svg>
          </button>
        </div>
        <div
          class="release-env-group-list"
        >
          <ul
            class="release-env-list"
          >
            <li
              class="env-card"
            >
              <div
                class="env-card-header"
              >
                <div
                  class="mdc-evolution-chip release-environment release-environment-upstream"
                  role="row"
                >
                  <span
                    class="mdc-evolution-chip__cell mdc-evolution-chip__cell--primary mdc-evolution-chip__action--primary"
                    role="gridcell"
                  >
                    <span
                      class="mdc-evolution-chip__text-name"
                    >
                      prod
                    </span>
                     
                    <span
                      class="mdc-evolution-chip__text-numbers"
                    />
                    <div
                      class="release-environment env-locks"
                    >
                      <div
                        aria-label="Lock Message: \\"envLock\\" | ID: \\"ui-envlock\\"  | Click to unlock. "
                        class=""
                        data-mui-internal-clone-element="true"
                      >
                        <button
                          aria-label=""
                          class="mdc-button button-lock"
                        >
                          <div
                            class="mdc-button__ripple"
                          />
                          <svg
                            class="env-card-env-lock-icon"
                            height="24px"
                            width="24px"
                          >
                            Locks_White.svg
                          </svg>
                        </button>
                      </div>
                    </div>
                  </span>
                </div>
                <div
                  class="env-card-app-locks"
                >
                  <div
                    aria-label="Lock Message: \\"appLock\\" | ID: \\"ui-applock\\"  | Click to unlock. "
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    <button
                      aria-label=""
                      class="mdc-button button-lock"
                    >
                      <div
                        class="mdc-button__ripple"
                      />
                      <svg
                        class="env-card-app-lock"
                      >
                        Locks.svg
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="content-area"
              >
                <div
                  class="content-left"
                >
                  <div
                    class="env-card-data"
                    title="Shows the version that is currently deployed on prod. "
                  >
                    commit: test1
                  </div>
                </div>
                <div
                  class="content-right"
                >
                  <div
                    class="env-card-buttons"
                  >
                    <button
                      aria-label="Add lock"
                      class="mdc-button env-card-add-lock-btn"
                    >
                      <div
                        class="mdc-button__ripple"
                      />
                      <svg
                        class="icon"
                      >
                        Locks.svg
                      </svg>
                      <span
                        class="mdc-button__label"
                      >
                        Add lock
                      </span>
                    </button>
                    <button
                      aria-label="Deploy"
                      class="mdc-button env-card-deploy-btn mdc-button--unelevated"
                      disabled=""
                    >
                      <div
                        class="mdc-button__ripple"
                      />
                      <span
                        class="mdc-button__label"
                      >
                        Deploy
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`Release Dialog Renders the environment locks two envs release 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div>
      <div />
    </div>
  </div>
  <div
    class="MuiModal-root MuiDialog-root release-dialog css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthMd MuiDialog-paperFullWidth css-142erj9-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <div
          class="release-dialog-app-bar"
        >
          <div
            class="release-dialog-app-bar-data"
          >
            <div
              class="release-dialog-message"
            >
              <span
                class="release-dialog-commitMessage"
              >
                test1
              </span>
            </div>
            <div
              class="release-dialog-createdAt"
            >
              <div>
                some formatted date
              </div>
            </div>
            <div
              class="release-dialog-author"
            >
              Author: test
            </div>
          </div>
          <span
            class="release-dialog-commitId"
            title="Commit Hash of the source repository."
          >
            commit
          </span>
          <button
            aria-label=""
            class="mdc-button release-dialog-close"
          >
            <div
              class="mdc-button__ripple"
            />
            <svg>
              Close.svg
            </svg>
          </button>
        </div>
        <div
          class="release-env-group-list"
        >
          <ul
            class="release-env-list"
          >
            <li
              class="env-card"
            >
              <div
                class="env-card-header"
              >
                <div
                  class="mdc-evolution-chip release-environment release-environment-upstream"
                  role="row"
                >
                  <span
                    class="mdc-evolution-chip__cell mdc-evolution-chip__cell--primary mdc-evolution-chip__action--primary"
                    role="gridcell"
                  >
                    <span
                      class="mdc-evolution-chip__text-name"
                    >
                      prod
                    </span>
                     
                    <span
                      class="mdc-evolution-chip__text-numbers"
                    />
                    <div
                      class="release-environment env-locks"
                    >
                      <div
                        aria-label="Lock Message: \\"envLock\\" | ID: \\"ui-envlock\\"  | Click to unlock. "
                        class=""
                        data-mui-internal-clone-element="true"
                      >
                        <button
                          aria-label=""
                          class="mdc-button button-lock"
                        >
                          <div
                            class="mdc-button__ripple"
                          />
                          <svg
                            class="env-card-env-lock-icon"
                            height="24px"
                            width="24px"
                          >
                            Locks_White.svg
                          </svg>
                        </button>
                      </div>
                    </div>
                  </span>
                </div>
                <div
                  class="env-card-app-locks"
                >
                  <div
                    aria-label="Lock Message: \\"appLock\\" | ID: \\"ui-applock\\"  | Click to unlock. "
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    <button
                      aria-label=""
                      class="mdc-button button-lock"
                    >
                      <div
                        class="mdc-button__ripple"
                      />
                      <svg
                        class="env-card-app-lock"
                      >
                        Locks.svg
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="content-area"
              >
                <div
                  class="content-left"
                >
                  <div
                    class="env-card-data"
                    title="Shows the version that is currently deployed on prod. "
                  >
                    commit: test1
                  </div>
                </div>
                <div
                  class="content-right"
                >
                  <div
                    class="env-card-buttons"
                  >
                    <button
                      aria-label="Add lock"
                      class="mdc-button env-card-add-lock-btn"
                    >
                      <div
                        class="mdc-button__ripple"
                      />
                      <svg
                        class="icon"
                      >
                        Locks.svg
                      </svg>
                      <span
                        class="mdc-button__label"
                      >
                        Add lock
                      </span>
                    </button>
                    <button
                      aria-label="Deploy"
                      class="mdc-button env-card-deploy-btn mdc-button--unelevated"
                      disabled=""
                    >
                      <div
                        class="mdc-button__ripple"
                      />
                      <span
                        class="mdc-button__label"
                      >
                        Deploy
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </li>
            <li
              class="env-card"
            >
              <div
                class="env-card-header"
              >
                <div
                  class="mdc-evolution-chip release-environment release-environment-upstream"
                  role="row"
                >
                  <span
                    class="mdc-evolution-chip__cell mdc-evolution-chip__cell--primary mdc-evolution-chip__action--primary"
                    role="gridcell"
                  >
                    <span
                      class="mdc-evolution-chip__text-name"
                    >
                      dev
                    </span>
                     
                    <span
                      class="mdc-evolution-chip__text-numbers"
                    />
                    <div
                      class="release-environment env-locks"
                    >
                      <div
                        aria-label="Lock Message: \\"envLock\\" | ID: \\"ui-envlock\\"  | Click to unlock. "
                        class=""
                        data-mui-internal-clone-element="true"
                      >
                        <button
                          aria-label=""
                          class="mdc-button button-lock"
                        >
                          <div
                            class="mdc-button__ripple"
                          />
                          <svg
                            class="env-card-env-lock-icon"
                            height="24px"
                            width="24px"
                          >
                            Locks_White.svg
                          </svg>
                        </button>
                      </div>
                    </div>
                  </span>
                </div>
                <div
                  class="env-card-app-locks"
                >
                  <div
                    aria-label="Lock Message: \\"appLock\\" | ID: \\"ui-applock\\"  | Click to unlock. "
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    <button
                      aria-label=""
                      class="mdc-button button-lock"
                    >
                      <div
                        class="mdc-button__ripple"
                      />
                      <svg
                        class="env-card-app-lock"
                      >
                        Locks.svg
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="content-area"
              >
                <div
                  class="content-left"
                >
                  <div
                    class="env-card-data"
                    title="Shows the version that is currently deployed on dev. "
                  >
                    cafe: the other commit message 3
                  </div>
                  <div
                    class="env-card-data env-card-data-queue"
                    title="An attempt was made to deploy version 666 either by a release train, or when a new version was created. However, there was a lock present at the time, so kuberpult did not deploy this version. "
                  >
                    Version 
                    666
                     was not deployed, because of a lock.
                  </div>
                </div>
                <div
                  class="content-right"
                >
                  <div
                    class="env-card-buttons"
                  >
                    <button
                      aria-label="Add lock"
                      class="mdc-button env-card-add-lock-btn"
                    >
                      <div
                        class="mdc-button__ripple"
                      />
                      <svg
                        class="icon"
                      >
                        Locks.svg
                      </svg>
                      <span
                        class="mdc-button__label"
                      >
                        Add lock
                      </span>
                    </button>
                    <button
                      aria-label="Deploy"
                      class="mdc-button env-card-deploy-btn mdc-button--unelevated"
                    >
                      <div
                        class="mdc-button__ripple"
                      />
                      <span
                        class="mdc-button__label"
                      >
                        Deploy
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
